### ISSUE TITLE
Unable to allocate memory with ReLERNN_TRAIN_POOL

### ISSUE THREAD

**Issue by tjthurman** [kr-colab/ReLERNN#24]
I'm running into memory issues with ReLERNN_TRAIN_POOL. I'm not sure if this is a ReLERNN problem, or (more likely) something about the way my cluster and GPUs are set up. 

I'm running my analysis on a cluster (CPU: Intel Xeon Gold 6240 @ 2.60GHz, GPU: NVIDIA RTX 2080Ti), using 24 threads. I installed the dependencies through conda, here are the versions I'm using:

tensorflow | 2.2.0
cudatoolkit     | 10.2.89
cudnn           | 7.6.5
msprime         | 0.7.4
scikit-learn    | 0.23.1
scikit-allel    | 1.3.1
matplotlib      | 3.2.2

While ReLERNN_TRAIN_POOL runs I get frequent warnings:

```
WARNING:tensorflow:multiprocessing can interact badly with TensorFlow, causing nondeterministic deadlocks. For high performance data pipelines tf.data is recommended.
```

It runs for a while and then eventually (~24 hours, 49 epochs) relernn gives a memory allocation error:

```
Traceback (most recent call last):
  File "/home/tt164677e/anaconda3/envs/tf/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/home/tt164677e/anaconda3/envs/tf/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/tt164677e/anaconda3/envs/tf/lib/python3.8/site-packages/tensorflow/python/keras/utils/data_utils.py", line 843, in _run
    with closing(self.executor_fn(_SHARED_SEQUENCES)) as executor:
  File "/home/tt164677e/anaconda3/envs/tf/lib/python3.8/site-packages/tensorflow/python/keras/utils/data_utils.py", line 820, in pool_fn
    pool = get_pool_class(True)(
  File "/home/tt164677e/anaconda3/envs/tf/lib/python3.8/multiprocessing/context.py", line 119, in Pool
    return Pool(processes, initializer, initargs, maxtasksperchild,
  File "/home/tt164677e/anaconda3/envs/tf/lib/python3.8/multiprocessing/pool.py", line 212, in __init__
    self._repopulate_pool()
  File "/home/tt164677e/anaconda3/envs/tf/lib/python3.8/multiprocessing/pool.py", line 303, in _repopulate_pool
    return self._repopulate_pool_static(self._ctx, self.Process,
  File "/home/tt164677e/anaconda3/envs/tf/lib/python3.8/multiprocessing/pool.py", line 326, in _repopulate_pool_static
    w.start()
  File "/home/tt164677e/anaconda3/envs/tf/lib/python3.8/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/home/tt164677e/anaconda3/envs/tf/lib/python3.8/multiprocessing/context.py", line 276, in _Popen
    return Popen(process_obj)
  File "/home/tt164677e/anaconda3/envs/tf/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/home/tt164677e/anaconda3/envs/tf/lib/python3.8/multiprocessing/popen_fork.py", line 70, in _launch
    self.pid = os.fork()
OSError: [Errno 12] Cannot allocate memory
```

This error doesn't cause any out-of-memory errors in the scheduler and doesn't halt the program: I just notice that the logs stop updating and I have to manually stop the job.

Here's the command I ran:

```
ReLERNN_TRAIN_POOL -d genome_MQ20_minDP90_maf05 --readDepth 255 --maf 0.05 -t 24
```

If it helps, this is coming from a pool of 81 diploid individuals (so I specified `--sampleDepth 162` in ReLERNN_SIMULATE_POOL). The genome size is ~3Gb, and I have just under 55k SNPs at the current level of filtering. The cluster I'm running on has I think up to 512 Gb of memory to work with, though if I check memory usage of the failed job with `sacct` I get some nonsensical numbers (MaxRSS = 18130.31G, MaxVMSize = 19035.73G), so I'm not sure what's going on there. 

Also if it helps, I was successfully able to run the example pooled pipeline, but it took longer than I was expecting given what the readme says for the non-pooled example: ~60 minutes running on 4 cores.

Please let me know if you need any other info. I'm very new with running GPU-based analyses, so even if you can just point me in the right direction in terms of questions to ask my sysadmin, I'd appreciate it.

**Comment by jradrion** [comment]
Hi @tjthurman, sorry for such a slow response. I completely lost track of this issue.

The first thing to address is making sure that training is occurring on the GPU and not the CPU. The example pipeline for pooled ReLERNN should not be taking an hour to run on an NVIDIA 2080, but more like ~5-6 minutes. Can you open a python interpreter, import tensorflow, and then run `tf.test.is_gpu_available()`. That command should ultimately return `True`, and if it's returning `False` you will likely need to reinstall tensorflow and cuda. I have noticed some weird incompatibilities between specific tensorflow and cuda versions. I can't remember if there was an issue with cudatoolkit 10.2, but I might try installing cudatoolkit/10.1.243 if possible. Uninstall tensorflow/cudatookit/cudnn, then reinstall cudatoolkit and cudnn before finally reinstalling tensorflow. If `tf.test.is_gpu_available()` is returning `True` yet still taking an hour to run the example pipeline, I'm not exactly sure what might be going on and would need you to send me the stdout and stderr from running that pipeline.

**Comment by tjthurman** [comment]
No worries, @jradrion, and thanks for the response!

I checked `tf.test.is_gpu_available()`, and indeed they weren't. So I guess something went wrong on install. I downgraded to cudatoolkit/10.1.243, re-installed tensorflow, and that seems to have done the trick: ` tf.test.is_gpu_available()` returned true and  `tf.config.list_physical_devices('GPU')` found 4 GPUs. I re-ran the example pipeline, and it only took 6 minutes. 

I'll give my actual pipeline another try and see if I run into the error again. Thanks for the help!
